### DISCUSSION TITLE
Reducing package size

### DISCUSSION THREAD

**Discussion by frankdejonge** [aws/aws-sdk-php#2420] [General]
### Describe the feature

It would be great if this library could publish sub-splits to publish smaller sub-packages. The current artefact is 29MB, a substantial amount of data. Many users only use a fraction of the total package size, yet the code is shipped over and over. As an example, the Symfony codebase uses sub-splits to allow independent distribution of components. Flysystem, the project I maintain, does the same for all of the adapters. I've personally set this up and would be willing to help out setting it up or answering questions about it. I believe this would be hugely beneficial for our users and also reduces the environmental impact.

Please consider this 🙏

### Use Case

When I download this package, I download 29MB while I use but a fraction of the source code.

### Proposed Solution

Setup splish/lite (https://github.com/splitsh/lite)

### Acknowledgements

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### SDK version used

Irrelevant

### Environment details (Version of PHP (`php -v`)? OS name and version, etc.)

Irrelevant

**Comment by stobrien89** [comment]
Hi @frankdejonge,

Thanks for the feature request.  We've given this serious consideration, but we're not in a position to manage sub-trees/repositories at this time.  We plan to make the SDK modular in the next major version and will likely use Satis or similar.  

That said, we realize package size is an issue for a number of customers and would like to know how many people would be interested in a [feature that cleans up unused services](https://github.com/aws/aws-sdk-php/issues/2369). We're open to other suggestions as well— I'm going to convert this to a discussion for now so we can get an idea of what path would be best for everyone.

**Reply by frankdejonge** [reply]
Hi @stobrien89 I think any effort to reduce package size is great. When it comes to the proposed solution, there are might be some unfortunate downsides. With containers being a common artefact form for distributing application, each command results in a layer, so people need to be very aware that these commands need to not suffer from having an intermediate layer where _does_ contain the code. That said, it does make it possible to reduce package size, which is great.

**Reply by tomasfejfar** [reply]
The feature to clean up unused stuff afterwards is really not ideal. One example being using Docker for Windows. Disk IO is rather slow in docker mounted volume and usual `composer install` call takes 10s to finish everything except aws sdk, which takes additional 100s or so. Also it's one more different layer of knowledge for the developers to maintain - when you need another AWS service, you need to know that there is this proprietary way of dealing with AWS SDK dependency. I sense many "I have aws-sdk-php required in composer, but S3 client cannot be found" bug reports. 

I think the split into multiple packages could be very crude with very basic dependencies you already have (to build the "cleanup tool") and still solve the problem completely. 
There could be some "composer dependency magic" to resolve the complexities into versioning the different packages. For example every package could require some `aws-sdk-php/package-lock` dummy package in a same version the package is released under.  

Example:

* `aws-sdk-php/s3:123.1.8` requires `aws-sdk-php/package-lock:123.1.8`
* `aws-sdk-php/sqs:123.1.8` requires `aws-sdk-php/package-lock:123.1.8`
* `aws-sdk-php/sns:130.0.0` requires `aws-sdk-php/package-lock:130.0.0`

This way you don't have to deal with compatibility between `aws-sdk-php/sns:130.0.0` and `aws-sdk-php/sqs:123.1.8`, because those have conflicting dependency on `aws-sdk-php/package-lock:130.0.0` and cannot be installed together.

**Reply by stobrien89** [reply]
Hi @tomasfejfar,

Thanks for the feedback.  We realize the Composer script isn't the ideal solution, but we wanted to offer a compromise since we've made the decision not to modularize/split this major version of the SDK.  Modularization is something we've committed to for the next major version.  That being said, we're still willing to entertain other ideas that may make this easier for folks in containerized environments.

**Comment by stobrien89** [comment]
Just to clarify, anyone who is in favor of the feature I mentioned above can upvote my first comment.  Anyone who is not in favor and/or has other ideas or suggestions is welcome to provide feedback.

**Comment by tomasfejfar** [comment]
Another case for reducing the package size is reducing the time to index the library in IDE (looking at you, PhpStorm). AWS SDK is by far the most released library and it takes by far the most time to index.

**Comment by stobrien89** [comment]
Hi all, 

Just wanted to mention we've just merged [a feature which allows you to remove unused services](https://github.com/aws/aws-sdk-php/tree/master/src/Script/Composer).  It should be available in tomorrow's release.  Any feedback is welcomed and appreciated.

**Reply by devbanana** [reply]
Oh this is a lifesaver, thank you for including this! Works great from my testing so far.

**Reply by longshadowdev** [reply]
Thank you for this new feature. It's something I've been very keen on, as I run php in AWS Lambda and package size is a key concern of mine always.

One thing I wanted to check is if this is recursive? For example I create a helper package that requires AWS SNS only. In my actual application repo I require AWS S3 and SecretsManager, plus my internal helper package that depends on AWS SNS.

If I specify SNS only in the helper package, and then S3 and SecretsManager in my app, will I preserve SNS, S3 and SecretsManager in my vendor dir? Or will it only have S3 and SecretsManager, therefore my app needs to scrutinise any dependencies for their usage of the AWS SDK?

Thanks

**Reply by stobrien89** [reply]
Hi @longshadowdev,

It does not recursively check for additional dependencies on the package, meaning you'll need to specify which services you'd like to keep at each level you're using the SDK.  

S3 is one service that is required by SDK core functionality, so it will be available to you whether you specify it or not.  It seems you would need to specify both SNS and SecretsManager in your app, but it's difficult to say without knowing the exact dependency mapping of your application.

**Reply by longshadowdev** [reply]
Thank you. Just to be clear though, this would mean that anyone who wrote a utility library intended to be used in many different applications would not be able to use this feature. For example, OP makes Flysystem (great package btw OP). If he was to implement this to limit it to only S3, and then I was to use Flysystem but I also wanted SecretsManager, the Flysystem use of this plugin would have already stripped SecretsManager out?

I can do a little POC but I think that's what would happen. Each dependency doesn't store its own versions of dependencies with composer, rather they will share the one dependency, and if they modify it they'll be modifying it globally. So you'd need to recurve down into all copies of composer.json and merge all of the extra/aws/aws-sdk-php arrays together before removing the files from the SDK.

Until then you'd need to say no one that makes a public utility should use this feature.

**Reply by tomasfejfar** [reply]
I don't think it works like you describe @longshadowdev. I don't think the hooks are executed for each dependency. 

Instead you as end user are responsible to put all the required packages in your root composer.json's extra section. If you don't put there all the required packages, it will fail. If you don't have the hook set up for your root repo, nothing will happen IMHO.

**Reply by lechatthecat** [reply]
Does this mean it's user's responsibility to check which service is depending on which service and put all of the required services in composer.json? users need to manually create such dependency mapping as long as users use the reducing package size feature?

**Reply by stobrien89** [reply]
@lechatthecat There are only a few services that the SDK's core functionality depends on.  As far as I can remember, there are not services that are dependent on other services.  All services that are unsafe to delete are marked as such and will not be deleted under any circumstance.

In terms of services that depend on one another in a customer's application architecture, it's entirely up to the customer to determine those dependencies and account for them in the services they specify to keep.

**Reply by lechatthecat** [reply]
@stobrien89 
Thank you for your answer!

**Reply by Tofandel** [reply]
Note: this is only a 1/3rd solution, it will still require the full download and then removal of the lib via composer, this is not optimal, uses a lot of SSD write cycles and a lot of bandwidth for no valid reason

A switch to a private packagist or https://packagist.org/packages/symplify/monorepo-builder that creates readonly github repos under a new org, would be a much better and future proof solution (of course this will require all current users to require one package per actual package under different names instead of the whole thing) and this can be done by deprecating the old package with an indication about it

**Comment by fredericgboutin-yapla** [comment]
Another case for reducing the package size is when using AWS Lambda over a custom PHP provider - for example with https://bref.sh/.

There is a limit to the size of the produced package which is something like 220MB. This vendor component alone accounts for almost 16% of that quota, which is huge.

The dumped documentation in `src/data` subfolders are entirely to blame here.

I suppose I'll give a shot to https://github.com/aws/aws-sdk-php/discussions/2420#discussioncomment-3087924 🤞

**Comment by sc0ttdav3y** [comment]
I'm running on bref.sh and have found the tool https://github.com/aws/aws-sdk-php/discussions/2420#discussioncomment-3087924 has been useful to take a 40 MB dependency down to 5 MB. Great job and thanks @stobrien89 and team.

**Comment by yenfryherrerafeliz** [comment]
Hello everyone, currently we are exploring making potential improvements in the AWS SDK for PHP. We have created [this](https://github.com/aws/aws-sdk-php/issues/3017) GitHub issue to better gather feedback from our community. Please look into the options there and vote for the things you think will add value to you, if present in the AWS SDK for PHP.

Thank you!

**Comment by Radiergummi** [comment]
What the hell. This is completely out of proportion:
```php
du -hd1 ./vendor/aws/aws-sdk-php | sort -h
 48K    ./vendor/aws/aws-sdk-php/.github
476K    ./vendor/aws/aws-sdk-php/features
3.4M    ./vendor/aws/aws-sdk-php/build
5.2M    ./vendor/aws/aws-sdk-php/tests
7.9M    ./vendor/aws/aws-sdk-php/.changes
235M    ./vendor/aws/aws-sdk-php/src
475M    ./vendor/aws/aws-sdk-php/.git
729M    ./vendor/aws/aws-sdk-php
```

**729 Megabytes!** Why is the entire git folder in there? And 235 Megabytes of source code?

**Reply by fredericgboutin-yapla** [reply]
I'm not sure how you endup with the `.git` folder. Normally through `composer` you would only have the `src` folder, basically

**Reply by Tofandel** [reply]
Post your composer.json file

**Reply by Radiergummi** [reply]
Yeah, I realised that too and did some investigating; looks like at some point, we had a DNS resolution error while downloading packages, and it defaulted to cloning from GitHub, and that made its way into the lockfile… just in case, I added a `rm -rf vendor/aws/aws-sdk-php/.git` to the post-autoloader-dump command nevertheless, so that won't happen anymore 😄
